src/devices/ikea_sparsnas.c

```diff
@@ -0,0 +1,305 @@
+    key[4] = (uint8_t)(sensor_id_sub >> 16);
+    
+    
+    for (size_t i = 0; i < 5; i++)
+        decrypted[i] = buffer[i];
+
+    for(size_t i = 0; i < 13; i++)
+        decrypted[5 + i] = buffer[5 + i] ^ key[i % 5];
+
+    // Additional integrity checks
+    uint32_t rcv_sensor_id = decrypted[5] << 24 | decrypted[6] << 16 | decrypted[7] << 8 | decrypted[8];
+
+    if (decoder->verbose > 1){
+        fprintf(stderr, "IKEA Sparsnäs: CRC OK (%X == %X)\n", crc_calculated, crc_received);
+        fprintf(stderr, "IKEA Sparsnäs: Encryption key: 0x%X%X%X%X%X\n", key[0], key[1], key[2], key[3], key[4]);
+        decoder_output_bitrowf(decoder, decrypted, 18 * 8, "Decrypted");
+        fprintf(stderr, "IKEA Sparsnäs: Received sensor id: %d\n", rcv_sensor_id);
+    }
+    
+    if (rcv_sensor_id != ikea_sparsnas_sensor_id) {
+        if (decoder->verbose > 1){
+            fprintf(stderr, "IKEA Sparsnäs: Malformed package, or wrong sensor id. Received sensor id (%d) not the same as sender (%d)\n", rcv_sensor_id, ikea_sparsnas_sensor_id);
+        }
+    }
+
+    if ((!ikea_sparsnas_sensor_id) || (rcv_sensor_id != ikea_sparsnas_sensor_id)){
+        
+        data_t *data;
+        data = data_make(
+            "model",         "Model",               DATA_STRING, "IKEA Sparsnäs Energy Meter Monitor [Encrypted]",
+            "id",            "Sensor ID",           DATA_INT, ikea_sparsnas_sensor_id,
+            "mic",           "Integrity",           DATA_STRING,    "CRC",
+            NULL
+        );
+        decoder_output_data(decoder, data);
+        return 1;
+    }
+
+    if (decrypted[0] != 0x11){
+        decoder_output_bitrowf(decoder, decrypted + 5, 13 * 8,  "Message malformed");
+        if (decoder->verbose > 1){
+            fprintf(stderr, "IKEA Sparsnäs: Message malformed (byte0=%X expected %X)\n", decrypted[0], 0x11);
+        }
+        return 0;
+    }
+    if (decrypted[3] != 0x07){
+        decoder_output_bitrowf(decoder, decrypted + 5, 13 * 8,  "Message malformed");
+        if (decoder->verbose > 1){
+            fprintf(stderr, "IKEA Sparsnäs: Message malformed (byte3=%X expected %X)\n", decrypted[0], 0x07);
+        }
+        return 0;
+    }
+
+    //Value extraction and interpretation
+    uint16_t sequence_number = (decrypted[9] << 8 | decrypted[10]);
+    uint16_t effect = (decrypted[11] <<  8 | decrypted[12]);
+    uint32_t pulses = (decrypted[13] << 24 | decrypted[14] << 16 | decrypted[15] << 8 | decrypted[16]);
+    uint8_t battery =  decrypted[17];
+    double watt = effect * 24.;
+    uint8_t mode = decrypted[4]^0x0f;
+
+    if(mode == 1){     //Note that mode cycles between 0-3 when you first put in the batteries in
+      watt = ((3600000.0 / ikea_sparsnas_pulses_per_kwh) * 1024.0) / effect;
+    } else if (mode == 0 ) { // special mode for low power usage
+      watt = effect * 0.24 / ikea_sparsnas_pulses_per_kwh;
+    }
+    double cumulative_kWh = ((double)pulses) / ((double)ikea_sparsnas_pulses_per_kwh);
```

Comment:
All doubles should be changed to float.

Reply:
Why so? I still need a double in the end to pass it in DATA_DOUBLE?

Reply:
The point is to use the type with enough precision for the job. float has 4 bytes of storage and double usually uses 8 so by using float we save a few bytes and on performance limited platforms (if there are any left) the double emulation is assumed to be more costly compared to the float version. DATA_DOUBLE should be changed to float.